File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/Salsa/Hardware/Motor.java
package org.firstinspires.ftc.teamcode.Salsa.Hardware;

import com.qualcomm.robotcore.eventloop.opmode.OpMode;
import com.qualcomm.robotcore.hardware.DcMotor;
import com.qualcomm.robotcore.hardware.DcMotorImpl;
import com.qualcomm.robotcore.hardware.HardwareMap;

/**
 * Created by adityamavalankar on 11/5/18.
 */

public abstract class Motor implements DcMotor {

    public HardwareMap hwmap;
    private DcMotor dcm;

    public void init(String hardwareName, HardwareMap inputMap) {

        hwmap = inputMap;

        this.dcm = hwmap.get(DcMotor.class, hardwareName);
    }

}

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/Components/Motor.kt
package org.firstinspires.ftc.teamcode.Components
import com.qualcomm.robotcore.hardware.DcMotor
import com.qualcomm.robotcore.hardware.HardwareMap
import org.firstinspires.ftc.teamcode.Utils.Logger

public class Motor(val hardwareMap: HardwareMap,val motorName:String ){
    val motor = hardwareMap.dcMotor.get(motorName)
    val l = Logger("MOTOR", motorName)

    fun setPower(power:Double){
        motor.power = power
    }

    fun useEncoder(){
        motor.mode = DcMotor.RunMode.RUN_USING_ENCODER
    }

    fun prepareEncoderDrive(){
        motor.mode=DcMotor.RunMode.STOP_AND_RESET_ENCODER
        useEncoder()
    }

    fun stop(coast:Boolean=false){
        if(coast){
            motor.zeroPowerBehavior = DcMotor.ZeroPowerBehavior.FLOAT
        }else{
            motor.zeroPowerBehavior = DcMotor.ZeroPowerBehavior.BRAKE
        }
        motor.power = 0.0
    }


    fun logInfo(){
        l.logData("Power",motor.power)
        l.logData("Position",motor.currentPosition)
        l.logData("Busy",motor.isBusy)
        l.logData("Zero Power Behaviour",motor.zeroPowerBehavior.toString())
    }

}